agenta-cli/agenta/sdk/config_manager.py

```diff
@@ -0,0 +1,205 @@
+            or ("variant" in context and context["variant"])
+        ):
+            raise ValueError(
+                "Either config, environment or variant must be provided. Not both."
+            )
+        if "config" in context and context["config"]:
+            return schema(**context["config"])
+        elif "environment" in context and context["environment"]:
+            return ConfigManager.get_from_backend(
+                schema, environment=context["environment"]
+            )
+        elif "variant" in context and context["variant"]:
+            return ConfigManager.get_from_backend(schema, variant=context["variant"])
+        else:
+            raise ValueError("Either config, environment or variant must be provided")
+
+    @staticmethod
+    def get_from_backend(
+        schema: Type[T],
+        environment: Optional[str] = None,
+        version: Optional[str] = None,
+        variant: Optional[str] = None,
+    ) -> T:
+        """
+        Pulls the parameters for the app variant from the server and returns a config object.
+
+        This method retrieves the configuration from the backend server based on the provided
+        environment or variant. It then validates and returns the configuration as an instance
+        of the specified schema.
+
+        Args:
+            schema (Type[T]): A Pydantic model class that defines the structure of the configuration.
+            environment (Optional[str]): The environment name to fetch the configuration for.
+                                         Must be one of "development", "production", or "staging".
+            version (Optional[str]): Currently not implemented. Will raise NotImplementedError if provided.
+            variant (Optional[str]): The variant name to fetch the configuration for.
+
+        Returns:
+            T: An instance of the specified schema populated with the configuration data.
+
+        Raises:
+            ValueError: If neither environment nor variant is provided.
+            NotImplementedError: If a specific version is requested (not yet implemented).
+            ValidationError: If the retrieved configuration data doesn't match the schema.
+            Exception: For any other errors during the process (e.g., API communication issues).
+
+        Note:
+            Either environment or variant must be provided, but not both.
+        """
+        if not ConfigManager.client:
+            try:
+                ConfigManager.client = AgentaApi(
+                    base_url=singleton.host + "/api",
+                    api_key=singleton.api_key if singleton.api_key else "",
+                )
+            except Exception as ex:
+                logger.error(
+                    "Failed to initialize Agenta client with error: %s", str(ex)
+                )
+                raise
+        if not environment and not variant:
+            raise ValueError("Either environment or variant must be provided")
+        try:
```

Comment:
So there is a precedence of `environment` > `variant` (> `version` ?).
Aren't those all concept that can apply at the same time ? Like,
- this `variant` at this `version`
- this `variant` from this `environment`
and so on.

Reply:
I don't think so. 

In `environments` you can only get a `version`. You cannot get a `variant` from an `environment`

So you have two choices. Are you fetching a revision from an environment, or are you fetching directly the variant (using id). 

In both cases, you might able to fetch a certain `version`.  However, I am still uncertain whether to move forward with `versions` for `variants` or deprecate this. Since it seems that it complicates things too much... That is why for now I did not implement `variant/version` and made it either:

- you fetch the latest version in an environment
- you fetch a named version in an environment
- you fetch a variant.


What do you think?

Reply:
Oooph the full scope of this discussion would have to happen offline.
As far as this PR is concerned, as long as the current behavior does affect futureproofness, I'm good.
I guess, as long as `you fetch a variant` means `you fetch the latest version of a variant`, then this lgtm.